rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+  void (*wait_for_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
+
+  /*** TIMER CALLBACKS ***/
+  // Creates TF_Timer. Allocates timer resources on the underlying platform and initializes its
+  // internals.
+  void (*create_timer)(SE_Device* executor, SE_Timer* timer, SE_TimerFns** timer_fns, TF_Status* status);
+
+  // Destroy timer and deallocates timer resources on the underlying platform.
+  void (*destroy_timer)(SE_Device* executor, SE_Timer timer, SE_TimerFns* timer_fns);
+
+  // Records a start event for an interval timer.
+  TF_BOOL (*start_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+
+  // Records a stop event for an interval timer.
+  TF_BOOL (*stop_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+  /*** MEMCPY CALLBACKS ***/
+  // Entrains a memcpy operation onto stream, with a host destination location
+  // host_dst and a device memory source, with target size size.
+  TF_BOOL (*memcpy_to_host)(
+      SE_Device* executor, SE_Stream stream,
+      void* host_dst,
+      const SE_DeviceMemoryBase* device_src,
+      uint64_t size);
+
+  // Entrains a memcpy operation onto stream, with a device destination location
+  // and a host memory source, with target size size
+
+  TF_BOOL (*memcpy_from_host)(
+      SE_Device* executor, SE_Stream stream,
+      SE_DeviceMemoryBase* device_dst,
+      const void* host_src, uint64_t size);
+
+  // Causes the host code to synchronously wait for operations entrained onto
+  // stream to complete. Effectively a join on the asynchronous device
+  // operations enqueued on the stream before this program point.
+  void (*block_host_until_done)(
+      SE_Device* executor, SE_Stream stream,
+      TF_Status* status);
+
+  // Synchronizes all activity occurring in the StreamExecutor's context (most
+  // likely a whole device).
+  TF_BOOL (*synchronize_all_activity)(SE_Device* executor);
+
+  // Obtains metadata about the underlying device.
+  void (*fill_device_description)(SE_Device* executor,
+                               SE_DeviceDescription* description,
+                               TF_Status* status);
+
+  // Entrains on a stream a user-specified function to be run on the host.
+  TF_BOOL (*host_callback)(SE_Device* executor, SE_Stream* stream,
+                     TF_StatusCallbackFn callback_fn, void* ctx);
+} SE_StreamExecutor;
+
+#define SE_STREAMEXECUTOR_STRUCT_SIZE TF_OFFSET_OF_END(SE_StreamExecutor, host_callback)
+
+typedef struct SE_PlatformParams {
+  size_t struct_size;
+  SE_PlatformId* id;
+  
+  // Callbacks for creating/destroying SE_Device.
+  SE_Device* (*create_device)(SE_Options* options, TF_Status* status);
+  void (*destroy_device)(SE_Device* device);
+  
+  // Callbacks for creating/destroying SE_StreamExecutor.
+  SE_StreamExecutor* (*create_stream_executor)(TF_Status* status);
+  void (*destroy_stream_executor)(SE_StreamExecutor* stream_executor);
```

Comment:
Does @yisitu 's comment about heap vs. stack allocation also apply here?

Reply:
It should apply as well and needs to be cleaned up.

The main motivating reason is, when TF inits struct_size, plugin will be able to inspect and reason about TF's view of SE_PlatformParams. That way, the plugin has the freedom to choose to wire up different functionality given TF's struct_size (if plugin wishes to).

Now in contrast, if plugin were to set SE_PlatformParams without any inputs from TF, then plugin will never have the chance to switch-case (wire up) different functionality based what it knows about on TF.

The approach is to have an API, where the plugin receives from TF _a pointer to a struct initialized by TF_.

* Plugin should not care which memory space (heap/stack/bss) the pointer is pointing to.

Reply:
For completeness, another way we could phrase this API is (pseudo-code):

```
#define SE_PLUGIN_VERSION 42

typedef struct {
  SE_Device* (*create_device)(SE_PluginHooks* self, .. params ..);
} SE_PluginHooks;

SE_PluginHooks* InitializeSEPlugin(int version);  // TF calls InitializeSEPlugin(SE_PLUGIN_VERSION);
```

This way the `InitializeSEPlugin` knows the interface TF was built against and can (for instance) return a different set of hooks via `SE_PluginHooks`.

Reply:
Yes, we were planning to switch to call a function instead of static registration (according to comment on this doc: https://github.com/tensorflow/community/pull/262). I will try to update the doc accordingly in next 1-2 hours.

Reply:
It is possible to complement what we have and add `InitalizeSEPlugin(version)`. Both techniques are not mutually exclusive.

FWIW the versioning requirement is _bidirectional_; plugin needs to know TF's view, and TF also needs to know plugin's, and this is achieved by one mechanism that is size of struct. The struct size approach makes it very meta and self-contained.

Reply:
> FWIW the versioning requirement is _bidirectional_;

That's true; what I suggested above is not sufficient.  We would also need a `GetSePluginVersion()` function that TF can use to figure out the version of the loaded plugin.

Reply:
It can be simplified to just a single API:
```
// API
typedef struct InitializeSEPlugin_Params {
  size_t struct_size;
  void* data;
  uint64_t version;  // in/out param
} InitializeSEPlugin_Params;

// Plugin
void InitializeSEPlugin(InitializeSEPlugin* params) {
  ReadTFVersion(params->version);
  params->version = GetPluginVersion();
}
```

Reply:
This is essentially what I have now in updated doc with one notable difference: I have 3 struct members for version. Are you proposing one version value to save space?

Reply:
@yisitu, Yi Situ, can you check my question above regarding version.

Based on my memory, there was no strong opinion in design review. We have create and destroy functions right now for `SP_Device` and `SP_StreamExecutor`. So, all is good?

Reply:
All good! Please resolve.